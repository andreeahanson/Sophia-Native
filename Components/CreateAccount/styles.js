import { StyleSheet } from 'react-native';
import theme from '../../theme';

export const styles = StyleSheet.create({
	container: {
		backgroundColor: theme.accentOne,
		flex: 1,
		width: '100%'
	},
	text: {
		fontSize: 25,
		fontFamily: theme.textMain,
		textAlign: 'center',
		margin: 10
	},
	message: {
		fontSize: 16,
		fontFamily: theme.textMain,
		color: theme.primary,
		margin: 10
	}
});
