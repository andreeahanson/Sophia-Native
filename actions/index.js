export const loadProfile = profile => ({
  type: "LOAD_PROFILE",
  profile
})

export const loadLists = lists => ({
  type: "LOAD_LISTS",
  lists
})

export const addList = newList => ({
  type: "ADD_LIST",
  newList
})

export const deleteList = listId => ({
  type: "DELETE_LIST",
  listId
})

export const loadTasks = (listId, tasks) => ({
  type: "LOAD_TASKS",
  listId,
  tasks
})

export const addTask = newTask => ({
  type: "ADD_TASK",
  newTask
})

export const deleteTask = taskId => ({
  type: "DELETE_TASK",
  taskId
})

export const editTask = (nameToChange, taskId) => ({
  type: "EDIT_TASK",
  nameToChange,
  taskId
})