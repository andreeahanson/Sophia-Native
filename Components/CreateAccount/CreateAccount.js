import React, { Component } from 'react';
import { connect } from 'react-redux';
import {
	View,
	Text,
	StyleSheet,
	TextInput,
	ScrollView,
	KeyboardAvoidingView,
} from 'react-native';
import { TouchableHighlight } from 'react-native-gesture-handler';
import theme from '../../theme';
// import { PropTypes } from 'prop-types';
import { logIn } from '../../actions';
import { logInClient} from '../../Utils/loginClient';
import { logInCaretaker} from '../../Utils/loginCaretaker';

const initialState = {
	user: '',
	username: '',
	password: '',
	password_confirmation: '',
	name: '',
	address: '',
	city: '',
	state: '',
	zip: '',
	email: '',
	phone: '',
	needs: [],
	allergies: [],
	diet: [],
	medications: [],
	abilities: []
};

export class CreateAccount extends Component {
	state = initialState;

	handleChange = (name, value) => {
		const multiResponseInputs = [ 'needs', 'allergies', 'diet', 'medications', 'abilities' ];
		if (multiResponseInputs.includes(name)) {
			value = value.split(',');
		}
		this.setState({
			[name]: value
		});
	};

	handleClientSubmit = () => {
		const {
			username,
			password,
			password_confirmation,
			name,
			address,
			city,
			state,
			zip,
			email,
			phone,
			needs,
			allergies,
			diet,
			medications
		} = this.state;

		const newClientProfile = {
			username,
			password,
			password_confirmation,
			name,
			street_address: address,
			city,
			state,
			zip,
			email,
			phone_number: phone,
			needs,
			allergies,
			diet_restrictions: diet,
			medications
		};
		this.postClient(newClientProfile)
	};

	handleCaretakerSubmit = () => {
		const { username, password, password_confirmation, name, email, phone, abilities } = this.state;

		const newCaretakerProfile = {
			username,
			password,
			password_confirmation,
			name,
			email,
			phone_number: phone,
			abilities
		};
		// this.postCaretaker(newCaretakerProfile)
	};

	postClient = async profile => {
		const options = {
			method: 'POST',
			headers: { 'Content-Type': 'application/json' },
			body: JSON.stringify(profile)
		};
		try {
			const response = await fetch('https://sophia-be.herokuapp.com/api/v1/clients/', options);
			const account = await response.json();
			// await logInClient(account.username, account.password)
			// console.log(user)
			// this.props.logIn(user)
				this.setState(initialState);
    	this.props.navigation.navigate('ClientProfile');
		} catch (error) {
      throw new Error(`failed to post profile: ${error.message}`);
		}
	};

	postCaretaker = async profile => {
		const options = {
			method: 'POST',
			headers: { 'Content-Type': 'application/json' },
			body: JSON.stringify(profile)
    };
    
		try {
			const response = await fetch('https://sophia-be.herokuapp.com/api/v1/caretakers/', options);
      const account = await response.json();
      await logInCaretaker(account.username, account.password)
		} catch (error) {
			throw new Error(`failed to post profile: ${error.message}`);
		}
  };
  

	renderClientInput = () => {
		return (
			<View>
				<TextInput
					style={styles.input}
					placeholder="Street Address"
					onChangeText={value => this.handleChange('address', value)}
					placeholderTextColor={theme.primary}
				/>
				<TextInput
					style={styles.input}
					placeholder="City"
					onChangeText={value => this.handleChange('city', value)}
					placeholderTextColor={theme.primary}
				/>
				<TextInput
					style={styles.input}
					placeholder="State"
					onChangeText={value => this.handleChange('state', value)}
					placeholderTextColor={theme.primary}
				/>
				<TextInput
					style={styles.input}
					placeholder="Zip Code"
					onChangeText={value => this.handleChange('zip', value)}
					placeholderTextColor={theme.primary}
				/>
				<Text style={styles.text}>Seperate multiple input values by commas</Text>
				<TextInput
					style={styles.input}
					placeholder="Caretaking Needs"
					onChangeText={value => this.handleChange('needs', value)}
					placeholderTextColor={theme.primary}
				/>
				<TextInput
					style={styles.input}
					placeholder="Allergies"
					onChangeText={value => this.handleChange('allergies', value)}
					placeholderTextColor={theme.primary}
				/>
				<TextInput
					style={styles.input}
					placeholder="Dietary Restrictions"
					onChangeText={value => this.handleChange('diet', value)}
					placeholderTextColor={theme.primary}
				/>
				<TextInput
					style={styles.input}
					placeholder="Medications"
					onChangeText={value => this.handleChange('medications', value)}
					placeholderTextColor={theme.primary}
				/>
			</View>
		);
	};

	renderCaretakerInput = () => {
		return (
			<View>
				<Text style={styles.text}>Seperate multiple input values by commas</Text>
				<TextInput
					style={styles.input}
					placeholder="Caretaking Abilities"
					onChangeText={value => this.handleChange('abilities', value)}
					placeholderTextColor={theme.primary}
				/>
			</View>
		);
	};

	renderClientBtn = () => {
		return (
			<View style={styles.routes}>
				<TouchableHighlight
					underlayColor={theme.accentTwo}
					accessibilityLabel="Tap me to create your client account."
					accessible={true}
					onPress={this.handleClientSubmit}
					style={styles.touchExpander}>
					<Text style={styles.registerButton}>Register {this.state.user === 'client' ? 'Client' : null}</Text>
				</TouchableHighlight>
			</View>
		);
	};

	renderCaretakerBtn = () => {
		return (
			<View style={styles.routes}>
				<TouchableHighlight
					underlayColor={theme.accentTwo}
					accessibilityLabel="Tap me to create your caretaker account."
					accessible={true}
					onPress={this.handleCaretakerSubmit}
					style={styles.touchExpander}>
					<Text style={styles.registerButton}>Register {this.state.user === 'caretaker' ? 'Caretaker' : null}</Text>
				</TouchableHighlight>
			</View>
		);
	};

	render () {
		return (
			<KeyboardAvoidingView style={styles.container} behavior="padding" enabled>
					<View>
						<View style={styles.headerContainer}>
							<Text style={styles.header} accessibilityLabel="Fill in the inputs to create an account">
								Create Account
							</Text>
						</View>
						<View style={styles.routes}>
							<TouchableHighlight
								underlayColor={theme.accentTwo}
								accessibilityLabel="Tap me to create your client account."
								accessible={true}
								onPress={() => this.setState({ user: 'client' })}
								style={styles.touchExpander}>
								<Text style={styles.button}>I'm a Client</Text>
							</TouchableHighlight>
						</View>
						<View style={styles.routes}>
							<TouchableHighlight
								underlayColor={theme.accentTwo}
								accessibilityLabel="Tap me to create your caretaker account."
								accessible={true}
								onPress={() => this.setState({ user: 'caretaker' })}
								style={styles.touchExpander}>
								<Text style={styles.button}>I'm a Caretaker</Text>
							</TouchableHighlight>
						</View>
						<Text style={styles.text}>Scroll to fill out the form below:</Text>
						<ScrollView style={styles.scrollContainer}>
						<TextInput
							style={styles.input}
							placeholder="Username"
							onChangeText={value => this.handleChange('username', value)}
							placeholderTextColor={theme.primary}
						/>
						<TextInput
							style={styles.input}
							placeholder="Password"
							onChangeText={value => this.handleChange('password', value)}
							placeholderTextColor={theme.primary}
						/>
							<TextInput
							style={styles.input}
							placeholder="Password"
							onChangeText={value => this.handleChange('password_confirmation', value)}
							placeholderTextColor={theme.primary}
						/>
						<TextInput
							style={styles.input}
							placeholder="Your Name"
							onChangeText={value => this.handleChange('name', value)}
							placeholderTextColor={theme.primary}
						/>
						<TextInput
							style={styles.input}
							placeholder="Email"
							onChangeText={value => this.handleChange('email', value)}
							placeholderTextColor={theme.primary}
						/>
						<TextInput
							style={styles.input}
							placeholder="Phone"
							onChangeText={value => this.handleChange('phone', value)}
							placeholderTextColor={theme.primary}
						/>
						{this.state.user === 'client' && this.renderClientInput()}
						{this.state.user === 'caretaker' && this.renderCaretakerInput()}
						</ScrollView>
						{this.state.user === 'client' ? this.renderClientBtn() : null}
						{this.state.user === 'caretaker' ? this.renderCaretakerBtn() : null}
					</View>
				 </KeyboardAvoidingView>
		);
	}
}

const mapDispatchToProps = dispatch => ({
  logIn:(user => dispatch(logIn(user)))
});

export default connect(null, mapDispatchToProps)(CreateAccount);

const styles = StyleSheet.create({
	container: {
		backgroundColor: theme.accentOne,
		alignItems: 'center',
		justifyContent: 'center',
		height: '100%',
	},
	headerContainer: {
		borderBottomColor: theme.primary,
		borderBottomWidth: StyleSheet.hairlineWidth,
		marginBottom: 10,
		width: '80%'
	},
	header: {
		fontSize: 30,
		fontFamily: 'Didot'
	},
	routes: {
		flexDirection: 'column',
		backgroundColor: theme.primary,
		width: '90%',
		height: '10%',
		borderRadius: 30,
		justifyContent: 'space-evenly',
		margin: 5
	},
	scrollContainer: {
		marginTop: 10
	},
	button: {
		color: theme.accentOne,
		fontSize: 35,
		fontFamily: 'Didot',
		textAlign: 'center',
		marginTop: 10,
		width: '80%'
	},
	text: {
		fontSize: 25,
		fontFamily: 'Didot',
		textAlign: 'center',
		margin: 10
	},
	input: {
		width: '90%',
		height: 80,
		fontSize: 25,
		fontFamily: 'Didot',
		paddingLeft: 5,
		marginTop: 10,
		backgroundColor: theme.accentThree,
		color: theme.accentTwo
	},
	touchExpander: {
		height: '90%',
		borderRadius: 30,
		width: '90%'
	},
	registerButton: {
		fontSize: 25,
		color: theme.accentOne,
		fontFamily: 'Didot',
		textAlign: 'center',
		marginTop: 10
	}
});

// CreateAccount.propTypes = {
// };
