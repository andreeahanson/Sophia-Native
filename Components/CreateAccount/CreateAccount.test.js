import React from "react";
import { CreateAccount } from "./CreateAccount";
import { shallow } from "enzyme";
import "react-native";
import { mapStateToProps, mapDispatchToProps } from "./CreateAccount";
import { logIn } from "../../actions";

jest.mock("react-native-gesture-handler", () => {
  return {};
});

jest.mock("../../Utils/apiCalls", () => ({
  postClient: jest.fn().mockImplementation(() => {
    return {
      id: 1,
      name: "mockUser"
    };
  })
}));

global.fetch = jest.fn().mockImplementation(() => {
  return Promise.resolve({
    ok: true,
    json: () => Promise.resolve(mockUser)
  });
});

let mockPostClient = jest.fn()
let mockLogIn = jest.fn()

let wrapper = shallow(<CreateAccount 
postClient = {mockPostClient}
logIn={mockLogIn}
navigation={{ navigate: jest.fn() }}
/>);

test("CreateAccount renders correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

it("should return a userAccount object", () => {
  const initialState = {};

  let mappedProps = mapStateToProps(initialState);

  expect(mappedProps).toEqual(initialState);
});

it("calls dispatch with an object when logIn action is called", () => {
  let initialState = {
    logIn: {}
  };

  const mockDispatch = jest.fn();
  const mockAction = logIn(initialState.userAccount);

  const mappedProps = mapDispatchToProps(mockDispatch);
  mappedProps.logIn(initialState.lists);

  expect(mockDispatch).toHaveBeenCalledWith(mockAction);
});

test("change the state of every element of the initial state when handleChange is invoked", () => {
  wrapper.state.name = "";
  let mockValue = "changed";

  let mockName = "name";

  wrapper.instance().handleChange(mockName, mockValue);

  expect(wrapper.state("name")).toEqual(mockValue);
});

it("should set the message state to a string that asks the user to type in all the input fields when handleClientSubmit is invoked", () => {
  wrapper.state.message = "";
  wrapper.state.username = "filled"
  wrapper.state.password = "filled"
  wrapper.state.password_confirmation = "filled"
  wrapper.state.name = ""
  wrapper.state.email = "filled"
  wrapper.state.phone = "filled"
  wrapper.state.address = "filled"
  wrapper.state.state = "filled"
  wrapper.state.city = "filled"
  wrapper.state.zip = "filled"
  wrapper.state.needs = "filled"
  wrapper.state.allergies = "filled"
  wrapper.state.diet = "filled"
  wrapper.state.medications = "filled"
  let mockMessage = "Please fill out all input fields";

  wrapper.instance().handleClientSubmit();

  expect(wrapper.state("message")).toEqual(mockMessage);
});



