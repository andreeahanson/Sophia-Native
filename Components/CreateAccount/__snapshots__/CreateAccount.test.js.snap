// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateAccount renders correctly 1`] = `ShallowWrapper {}`;
